Improve install, usage, and configuration guide (#2677)
nf-core/website: sites/docs/src/content/docs/usage/configuration/modify_tools.md

---

> title: Tool arguments

Reviewer: This is a very confusing title in my opinion...

This is more something like 'Modifying pipelines' or 'Running unsupported functionality' or something..

Tool arguments coudl be interpreted simply how to modify a pipeline parameter with `--my_param` or something...

Reviewer: And has nothing to do with containers itself really

subtitle: Configure tool containers and arguments
shortTitle: Tool arguments
weight: 3
---

## Modifying tools

Each tool in an nf-core pipeline come preconfigured with a set arguments for an average user.

> The arguments are a great place to start and have been tested as a part of the development process.

Reviewer: ```suggestion
The arguments are a great place to start and have been tested as a part of the development process.
You normally can change the default settings using parameters using the double dash notation, e.g. `--input` or `--run_bwa`.
```

> However, you may want to modify these to fit your own purposes.

Reviewer: ```suggestion
However, you may want to use other versions of tools, or additional (pipeline unsupported) parameters these to fit your own purposes.
```

> **It is unlikely that you will need to edit the pipeline code to configure a tool.**

Reviewer: ```suggestion
**It is very unlikely that you will need to edit the pipeline code to configure a tool.**
```

> ### Docker registries

Reviewer: ```suggestion
### Changing source of container registries
```

I think this section should go later, it is much less rare that someone do this..

Author: Moved

nf-core pipelines use `quay.io` as the default docker registry for Docker and Podman images.
When specifying a Docker container, it will pull the image from `quay.io` unless a full URI is specified. For example, if the process container is:

```bash
biocontainers/fastqc:0.11.7--4
```

The image will be pulled from quay.io by default, resulting in a full URI of:

```bash
quay.io/biocontainers/fastqc:0.11.7--4
```

If `docker.registry` is specified, it will be used first. For example, if the config value `docker.registry = 'public.ecr.aws'` is specified the image will be pulled from:

```bash
public.ecr.aws/biocontainers/fastqc:0.11.7--4
```

However, the `docker.registry` setting will be ignored if you specify a full URI:

```bash
docker.io/biocontainers/fastqc:v0.11.9_cv8
```

> ### Tool versions

Reviewer: ```suggestion
### Changing tool versions
```

> The [Nextflow DSL2](https://www.nextflow.io/docs/latest/dsl2.html) implementation of nf-core pipelines uses one container or conda environment per process.

Reviewer: ```suggestion
```

> If you need to use a different version of a tool within a pipeline you need to identify the `process` name and override the Nextflow `container` or `conda` definition using the `withName` declaration.

Reviewer: ```suggestion
You can tell the pipeline to use a different container image within a config file and the `process` scope.

You then need to identify the `process` name and override the Nextflow `container` or `conda` definition using the `withName` declaration.
```

> For example, the [nf-core/viralrecon](https://nf-co.re/viralrecon) pipeline uses a tool called Pangolin that updates an internal database of COVID-19 lineages quite frequently. It doesn't make sense to re-release the `nf-core/viralrecon` pipeline every time a new version of Pangolin is released.

Reviewer: ```suggestion
For example, the [nf-core/viralrecon](https://nf-co.re/viralrecon) pipeline uses a tool called Pangolin that updates an internal database of COVID-19 lineages quite frequently.
```

> A user can override the default container used by the pipeline by creating a custom config file and passing it as a command-line argument via a custom configuration file.

Reviewer: ```suggestion
To update the container specification, you can do the following steps:
```

2. Find the latest version of the Biocontainer available on [quay.io](https://quay.io/repository/biocontainers/pangolin?tag=latest&tab=tags) for Docker or [Galaxy Project](https://depot.galaxyproject.org/singularity/) for Singularity
   - Note the container version tag is identical for both container systems, but must include the 'build' ID (e.g.`--pyhdfd78af_1`)
3. Create the custom config accordingly:

   - For Docker:

     ```groovy
     process {
         withName: PANGOLIN {
             container = 'quay.io/biocontainers/pangolin:3.1.17--pyhdfd78af_1'
         }
     }
     ```

   - For Singularity:

     ```groovy
     process {
         withName: PANGOLIN {
             container = 'https://depot.galaxyproject.org/singularity/pangolin:3.1.17--pyhdfd78af_1'
         }
     }
     ```

   - For Conda:

     ```groovy
     process {
         withName: PANGOLIN {
             conda = 'bioconda::pangolin=3.1.17'
         }
     }
     ```

:::warning

> Updated tools may come with major changes and may break a pipeline and/or create missing values in MultiQC version tables.

Reviewer: ```suggestion
Updated tools may come with major changes and may break a pipeline and/or create missing values in MultiQC version tables.

Such changes come with no warranty or support by the the pipeline developers! 
```

:::

> ### Tool arguments

Reviewer: This section should come first IMO, it's more likely to be requested

Author: Done

> You may wish to understand or update which tool arguments a pipeline uses.

Reviewer: ```suggestion
You may wish to understand which tool arguments a pipeline uses, or update to add additional arguments not supported currently by the pipeline. 
```

You can sometimes find out what parameters are used in a tool in by checking the longer 'help' description of different pipeline parameters, e.g., by pressing the 'help' button next to [this parameter](https://nf-co.re/funcscan/1.0.1/parameters#annotation_bakta_mincontig) in [nf-core/funcscan](https://nf-co.re/funcscan).

> There are two main places that a tool can have a tool argument specified:

Reviewer: ```suggestion
If not described there, there are two main places in the pipeline code that a tool can have a tool argument specified:
```

---
title: Tool arguments
subtitle: Configure tool containers and arguments
shortTitle: Tool arguments
weight: 3
---

## Modifying tools

Each tool in an nf-core pipeline come preconfigured with a set arguments for an average user.
The arguments are a great place to start and have been tested as a part of the development process.
However, you may want to modify these to fit your own purposes.

**It is unlikely that you will need to edit the pipeline code to configure a tool.**

### Docker registries

nf-core pipelines use `quay.io` as the default docker registry for Docker and Podman images.
When specifying a Docker container, it will pull the image from `quay.io` unless a full URI is specified. For example, if the process container is:

> `````bash
> biocontainers/fastqc:0.11.7--4
> ```

Reviewer: ```suggestion
  ```bash
  biocontainers/fastqc:0.11.7--4
  ```
```
